src/AIAPI/MainGame/GameCustomFunctionController.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+// using ActionGame;
+using ExtensibleSaveFormat;
+using KKAPI.Maker;
+using UnityEngine;
+using Manager;
+using AIProject;
+using AIChara;
+using CharaCustom;
+
+namespace KKAPI.MainGame
+{
+    /// <summary>
+    /// Base type for custom game extensions.
+    /// It provides many useful methods that abstract away the nasty hooks needed to figure out when the state of the game
+    /// changes.
+    /// 
+    /// This controller is a MonoBehaviour that is created upon registration in <see cref="GameAPI.RegisterExtraBehaviour{T}"/>.
+    /// The controller is created only once. If it's created too late it might miss some events.
+    /// It's recommended to register controllers in your Start method.
+    /// </summary>
+    public abstract class GameCustomFunctionController : MonoBehaviour
+    {
+        private static EnvironmentSimulator _environmentSimulator;
+
+        /// <summary>
+        /// Extended save ID used by this function controller
+        /// </summary>
+        public string ExtendedDataId { get; internal set; }
+        
+        /// <summary>
+        /// Get extended data based on supplied ExtendedDataId. When in chara maker loads data from character that's being loaded. 
+        /// </summary>
+        public PluginData GetExtendedData()
+        {
+            if (ExtendedDataId == null) throw new ArgumentException(nameof(ExtendedDataId));
+            return ExtendedSave.GetExtendedDataById(Singleton<Map>.Instance.Player.ChaControl.chaFile, ExtendedDataId);
```

Comment:
Clever, but isn't this going to cause issues if player character is changed, or if the plugin uses same ID for character and game controllers?

The rest looks perfectly fine.

Reply:
Huh for KK the first param is of type SaveData saveData, but for AI its ChaFile.  Maybe ExtensibleSaveFormat needs to be updated for AI?

Reply:
KK_ExtensibleSaveFormat has a [set of extra methods for main game saves](https://github.com/IllusionMods/BepisPlugins/blob/master/src/KK_ExtensibleSaveFormat/KK.ExtendedSave.SaveData.cs).  I think these need to be added for AI to support this.

Reply:
These methods can be stubbed for the time being (if you don't need them) with `throw new NotImplementedException()` and an `Obsolete` attribute with an explanation.

Reply:
Yea, I don't need them for now.  Ill Obsolete them for now.  If someone needs them later we can upgrade ExtensibleSaveFormat.